pkg/river/vm/op_binary.go

```diff
@@ -0,0 +1,324 @@
+package vm
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+
+	"github.com/grafana/agent/pkg/river/internal/value"
+	"github.com/grafana/agent/pkg/river/token"
+)
+
+func evalBinop(lhs value.Value, op token.Token, rhs value.Value) (value.Value, error) {
+	// We have special handling for EQ and NEQ since it's valid to attempt to
+	// compare values of any two types.
+	switch op {
+	case token.EQ:
+		return value.Bool(valuesEqual(lhs, rhs)), nil
+	case token.NEQ:
+		return value.Bool(!valuesEqual(lhs, rhs)), nil
+	}
+
+	// The type of lhs and rhs must be acceptable for the binary operator.
+	if !acceptableBinopType(lhs, op) {
+		return value.Null, value.Error{
+			Value: lhs,
+			Inner: fmt.Errorf("should be one of %s for binop %s, got %s", binopAllowedTypes[op], op, lhs.Type()),
+		}
+	} else if !acceptableBinopType(rhs, op) {
+		return value.Null, value.Error{
+			Value: rhs,
+			Inner: fmt.Errorf("should be one of %s for binop %s, got %s", binopAllowedTypes[op], op, rhs.Type()),
+		}
+	}
+
+	// At this point, regardless of the operator, lhs and rhs must have the same
+	// type.
+	if lhs.Type() != rhs.Type() {
+		return value.Null, value.TypeError{Value: rhs, Expected: lhs.Type()}
+	}
+
+	switch op {
+	case token.OR: // bool || bool
+		return value.Bool(lhs.Bool() || rhs.Bool()), nil
+	case token.AND: // bool && Bool
+		return value.Bool(lhs.Bool() && rhs.Bool()), nil
+
+	case token.ADD: // number + number, string + string
+		if lhs.Type() == value.TypeString {
+			return value.String(lhs.Text() + rhs.Text()), nil
+		}
+
+		lhsNum, rhsNum := lhs.Number(), rhs.Number()
+		switch fitNumberKinds(lhsNum.Kind(), rhsNum.Kind()) {
+		case value.NumberKindUint:
+			return value.Uint(lhsNum.Uint() + rhsNum.Uint()), nil
+		case value.NumberKindInt:
+			return value.Int(lhsNum.Int() + rhsNum.Int()), nil
+		case value.NumberKindFloat:
+			return value.Float(lhsNum.Float() + rhsNum.Float()), nil
+		}
+
+	case token.SUB: // number - number
+		lhsNum, rhsNum := lhs.Number(), rhs.Number()
+		switch fitNumberKinds(lhsNum.Kind(), rhsNum.Kind()) {
+		case value.NumberKindUint:
+			return value.Uint(lhsNum.Uint() - rhsNum.Uint()), nil
+		case value.NumberKindInt:
+			return value.Int(lhsNum.Int() - rhsNum.Int()), nil
+		case value.NumberKindFloat:
+			return value.Float(lhsNum.Float() - rhsNum.Float()), nil
+		}
```

Comment:
I would love to make these less repetitive but I haven't found a way to do that 😞